File: cardgrader/session.py
"""A grading session: the front of the card, the back, and the verdict.

Graders state a centering tolerance for each face and a card has to satisfy
both, so a front-only measurement can only ever be half the answer.  A session
holds an independent document per side -- its own scan, rotation, eight lines
and undo history -- and combines the two into one final ceiling.
"""

from __future__ import annotations

import json
import os
from datetime import datetime

from . import grades
from .grades import BACK, FRONT
from .model import CardModel

SIDES = (FRONT, BACK)
LABELS = {FRONT: "Front", BACK: "Back"}


class Session:
    def __init__(self):
        self.cards = {side: CardModel() for side in SIDES}
        self.active = FRONT

    # --- the side being worked on ----------------------------------------
    @property
    def card(self) -> CardModel:
        return self.cards[self.active]

    def set_active(self, side: str) -> bool:
        if side not in SIDES or side == self.active:
            return False
        self.active = side
        return True

    def other_side(self) -> str:
        return BACK if self.active == FRONT else FRONT

    def is_loaded(self, side: str) -> bool:
        return self.cards[side].loaded

    def loaded_sides(self):
        return [s for s in SIDES if self.cards[s].loaded]

    @property
    def any_loaded(self) -> bool:
        return bool(self.loaded_sides())

    @property
    def both_loaded(self) -> bool:
        return len(self.loaded_sides()) == 2

    # --- results ----------------------------------------------------------
    def measurement(self, side: str):
        card = self.cards[side]
        return card.measure() if card.loaded else None

    def side_ceilings(self, side: str):
        return grades.ceilings(self.measurement(side), side)

    def final_ceilings(self):
        """The card is held to whichever face reads worse."""
        return grades.combine(self.measurement(FRONT), self.measurement(BACK))

    def limiting_side(self):
        """Which face is holding the grade down, when they disagree."""
        front = {c.grader: c for c in self.side_ceilings(FRONT)}
        back = {c.grader: c for c in self.side_ceilings(BACK)}
        if not front or not back:
            return None
        worse_front = sum(1 for g in front if front[g].rank > back[g].rank)
        worse_back = sum(1 for g in back if back[g].rank > front[g].rank)
        if worse_front and not worse_back:
            return FRONT
        if worse_back and not worse_front:
            return BACK
        return None

    # --- reporting --------------------------------------------------------
    def report(self) -> dict:
        data = {
            "measured_at": datetime.now().isoformat(timespec="seconds"),
            "sides_measured": self.loaded_sides(),
            "sides": {},
        }
        for side in SIDES:
            card = self.cards[side]
            if card.loaded:
                data["sides"][side] = card.report(side)
        finals = self.final_ceilings()
        if finals:
            data["final_grade_ceilings"] = {
                c.grader: {"grade": c.grade, "descriptor": c.descriptor,
                           "set_by_side": c.side}
                for c in finals}
            data["limiting_side"] = self.limiting_side()
            data["grade_note"] = (
                "Centering only, and a ceiling rather than a prediction. "
                "Corners, edges and surface are not assessed, so the actual "
                "grade can be lower.")
            if not self.both_loaded:
                data["grade_caveat"] = (
                    "Only the %s was measured; the other face could pull the "
                    "grade down further." % ", ".join(self.loaded_sides()))
        return data

    def save_report(self, path: str) -> None:
        with open(path, "w", encoding="utf-8") as fh:
            json.dump(self.report(), fh, indent=2)

    # --- saving and reopening a session -----------------------------------
    def save_session(self, path: str) -> None:
        payload = {"version": 2, "active": self.active, "sides": {}}
        for side in SIDES:
            card = self.cards[side]
            if not card.loaded:
                continue
            payload["sides"][side] = {
                "path": card.path,
                "n_samples": card.n_samples,
                "transform": {"angle": card.transform.angle,
                              "scale": card.transform.scale,
                              "pan": list(map(float, card.transform.pan)),
                              "center": list(map(float, card.transform.center))},
                "lines": {k: [list(map(float, v.p1)), list(map(float, v.p2))]
                          for k, v in card.lines.items()},
            }
        with open(path, "w", encoding="utf-8") as fh:
            json.dump(payload, fh, indent=2)

    def load_session(self, path: str, canvas_size=(900, 700)):
        """Returns a list of sides whose image file could not be found."""
        with open(path, "r", encoding="utf-8") as fh:
            payload = json.load(fh)
        missing = []
        for side in SIDES:
            blob = payload.get("sides", {}).get(side)
            card = CardModel()
            if blob:
                img = blob.get("path")
                if img and os.path.exists(img):
                    card.load(img, canvas_size, auto=False)
                    card.apply_state(blob)
                else:
                    missing.append(side)
            self.cards[side] = card
        self.active = payload.get("active", FRONT)
        if not self.cards[self.active].loaded and self.any_loaded:
            self.active = self.loaded_sides()[0]
        return missing

File: src/core/session.ts
/**
 * A grading session: the front of the card, the back, and the verdict.
 *
 * Graders state a centering tolerance for each face and a card has to satisfy
 * both, so a session holds an independent document per side and combines the
 * two into one final ceiling.
 */

import type { Vec } from "./geometry";
import { type Ceiling, type Side, BACK, FRONT, GRADERS, ceilings, combine } from "./grades";
import type { Measurement } from "./measure";
import { type CardState, CardModel } from "./model";
import type { Raster } from "./raster";

export const SIDE_ORDER: readonly Side[] = [FRONT, BACK];
export const LABELS: Record<Side, string> = { front: "Front", back: "Back" };

export const SESSION_FORMAT = "card-centering-grader/session";
export const SESSION_VERSION = 3;

export interface SavedSide extends CardState { image: string }
export interface SavedSession {
  format: typeof SESSION_FORMAT;
  version: number;
  active: Side;
  sides: Partial<Record<Side, SavedSide>>;
}

export class Session {
  cards: Record<Side, CardModel> = { front: new CardModel(), back: new CardModel() };
  active: Side = FRONT;

  get card(): CardModel { return this.cards[this.active]; }

  setActive(side: Side): boolean {
    if (side === this.active) return false;
    this.active = side;
    return true;
  }

  otherSide(): Side { return this.active === FRONT ? BACK : FRONT; }
  isLoaded(side: Side): boolean { return this.cards[side].loaded; }
  loadedSides(): Side[] { return SIDE_ORDER.filter((s) => this.cards[s].loaded); }
  get anyLoaded(): boolean { return this.loadedSides().length > 0; }
  get bothLoaded(): boolean { return this.loadedSides().length === 2; }

  measurement(side: Side): Measurement | null {
    const c = this.cards[side];
    return c.loaded ? c.measure() : null;
  }

  sideCeilings(side: Side): Ceiling[] { return ceilings(this.measurement(side), side); }

  /** The card is held to whichever face reads worse. */
  finalCeilings(): Ceiling[] { return combine(this.measurement(FRONT), this.measurement(BACK)); }

  /** Which face is holding the grade down, when the two disagree. */
  limitingSide(): Side | null {
    const f = new Map(this.sideCeilings(FRONT).map((c) => [c.grader, c]));
    const b = new Map(this.sideCeilings(BACK).map((c) => [c.grader, c]));
    if (!f.size || !b.size) return null;
    let worseFront = 0;
    let worseBack = 0;
    for (const g of GRADERS) {
      if (f.get(g)!.rank > b.get(g)!.rank) worseFront++;
      if (b.get(g)!.rank > f.get(g)!.rank) worseBack++;
    }
    if (worseFront && !worseBack) return FRONT;
    if (worseBack && !worseFront) return BACK;
    return null;
  }

  report(now = new Date()): Record<string, unknown> {
    const data: Record<string, unknown> = {
      measured_at: now.toISOString().slice(0, 19),
      sides_measured: this.loadedSides(),
      sides: Object.fromEntries(this.loadedSides().map((s) => [s, this.cards[s].report(s)])),
    };
    const finals = this.finalCeilings();
    if (finals.length) {
      data.final_grade_ceilings = Object.fromEntries(finals.map((c) =>
        [c.grader, { grade: c.grade, descriptor: c.descriptor, set_by_side: c.side }]));
      data.limiting_side = this.limitingSide();
      data.grade_note = "Centering only, and a ceiling rather than a prediction. Corners, "
        + "edges and surface are not assessed, so the actual grade can be lower.";
      if (!this.bothLoaded) {
        data.grade_caveat = `Only the ${this.loadedSides().join(", ")} was measured; the other `
          + "face could pull the grade down further.";
      }
    }
    return data;
  }

  /** Serialise, embedding each source image so the file stands on its own. */
  async save(encode: (r: Raster) => Promise<string>): Promise<SavedSession> {
    const sides: Partial<Record<Side, SavedSide>> = {};
    for (const s of this.loadedSides()) {
      const card = this.cards[s];
      sides[s] = { ...card.state(), image: await encode(card.source!) };
    }
    return { format: SESSION_FORMAT, version: SESSION_VERSION, active: this.active, sides };
  }

  static async open(
    saved: SavedSession, decode: (s: string) => Promise<Raster>, canvasSize: Vec,
  ): Promise<Session> {
    if (saved?.format !== SESSION_FORMAT) throw new Error("This is not a Card Centering Grader session file.");
    const session = new Session();
    for (const s of SIDE_ORDER) {
      const blob = saved.sides?.[s];
      if (blob) session.cards[s].restore(await decode(blob.image), blob, canvasSize);
    }
    session.active = saved.active === BACK ? BACK : FRONT;
    if (!session.card.loaded && session.anyLoaded) session.active = session.loadedSides()[0];
    return session;
  }
}
